ast/desugar/Desugar.cc

```diff
@@ -1014,7 +1014,18 @@ ExpressionPtr node2TreeImpl(DesugarContext dctx, unique_ptr<parser::Node> what)
                     auto wrapped = MK::InsSeq(loc, std::move(stats), std::move(iff));
                     result = std::move(wrapped);
                 } else if (isa_reference(recv)) {
-                    auto cond = MK::cpRef(recv);
+                    ast::ExpressionPtr cond;
+                    auto *ident = cast_tree<UnresolvedIdent>(recv);
+                    if (ident && ident->kind == UnresolvedIdent::Kind::Class) {
+                        // Testing an uninitialized class variable directly will raise
+                        // an error, so we have to be more careful.
+                        auto sym = MK::Symbol(loc, ident->name);
+                        auto res = MK::Send1(loc, MK::Constant(loc, core::Symbols::Magic()),
+                                             core::Names::definedClassVar(), std::move(sym));
+                        cond = std::move(res);
+                    } else {
+                        cond = MK::cpRef(recv);
+                    }
```

Comment:
That case should be unaffected because this transform only kicks in for `@@var ||= ...`

Reply:
Hmm, how about this one?

https://sorbet.run/#%23%20typed%3A%20true%0A%0Aextend%20T%3A%3ASig%0A%0Aclass%20C%0A%20%20%40%40x%20%3D%20T.let%2842%2C%20Integer%29%0A%0A%20%20def%20f%0A%20%20%20%20%40%40x%20%7C%7C%3D%2043%0A%20%20end%0Aend%0A

Reply:
I believe (not having run this) that it would be affected, and we'd no longer complain about unreachable code there.  That seems like a totally reasonable thing to catch!

I think we could (partially) deal with that by writing a type deriver (I don't think that's the official name, but that's what I think of them as) in `calls.cc` for `<Magic>.<defined-class-var>`?  I'd have to experiment with that.

Reply:
Yeah, something like that sounds good! If we want to push ahead without that I'd probably want to solicit a little feedback from the team on whether we can live with the changed behavior around `@@non_nilable_thing ||=`.